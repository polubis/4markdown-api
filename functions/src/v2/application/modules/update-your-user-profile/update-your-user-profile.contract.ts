import { type UserProfileModel } from '@domain/models/user-profile';
import { base64, date, text, url } from '@utils/validators';
import { z } from 'zod';

const updateYourUserProfilePayloadSchema = z.object({
  mdate: date,
  displayName: text
    .min(2, `Display name must be at least 2 characters long`)
    .max(30, `Display name can be up to 30 characters long`)
    .nullable(),
  avatar: z.union([
    z.object({
      type: z.literal(`noop`),
    }),
    z.object({
      type: z.literal(`update`),
      data: base64(`Avatar must be in base64 format`),
    }),
    z.object({
      type: z.literal(`remove`),
    }),
  ]),
  bio: text
    .min(20, `Bio must be at least 20 characters long`)
    .max(500, `Bio can be up to 500 characters long`)
    .nullable(),
  githubUrl: url(`The GitHub URL format is invalid`).nullable(),
  fbUrl: url(`The Facebook URL format is invalid`).nullable(),
  linkedInUrl: url(`The LinkedIn URL format is invalid`).nullable(),
  blogUrl: url(`The website URL format is invalid`).nullable(),
  twitterUrl: url(`The Twitter URL format is invalid`).nullable(),
});

type UpdateYourUserProfilePayload = z.infer<
  typeof updateYourUserProfilePayloadSchema
>;
type UpdateYourUserProfileDto = {
  mdate: UserProfileModel['mdate'];
  profile: UserProfileModel;
};

export { updateYourUserProfilePayloadSchema };
export type { UpdateYourUserProfilePayload, UpdateYourUserProfileDto };
