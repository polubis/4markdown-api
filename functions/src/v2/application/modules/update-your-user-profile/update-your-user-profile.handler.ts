import { type ProtectedControllerHandlerContext } from '@utils/controller';
import {
  type UpdateYourUserProfileDto,
  type UpdateYourUserProfilePayload,
} from './update-your-user-profile.contract';
import { errors } from '@utils/errors';
import { type Transaction } from 'firebase-admin/firestore';
import { createSlug } from '@utils/create-slug';
import { type UserProfileModel } from '@domain/models/user-profile';
import { nowISO, uuid } from '@libs/helpers/stamps';
import { type Base64 } from '@utils/validators';
import { decodeBase64Asset } from '@utils/decode-base64-asset';
import { storage } from 'firebase-admin';
import * as sharp from 'sharp';

type UpdateYourUserProfileHandlerConfig = {
  payload: UpdateYourUserProfilePayload;
  context: ProtectedControllerHandlerContext;
};

const avatarVariants = [
  {
    size: `lg`,
    h: 100,
    w: 100,
  },
  {
    size: `md`,
    h: 64,
    w: 64,
  },
  {
    size: `sm`,
    h: 32,
    w: 32,
  },
  {
    size: `tn`,
    h: 24,
    w: 24,
  },
] as const;

const verifyDisplayNamesDuplication = async ({
  transaction,
  context,
  payload,
}: UpdateYourUserProfileHandlerConfig & {
  transaction: Transaction;
}): Promise<void> => {
  if (payload.profile.displayName === null) return;

  const userDisplayNamesSnap = await transaction.get(
    context.db
      .collection(`user-display-names`)
      .where(`__name__`, `==`, payload.profile.displayName)
      .where(`userId`, `!=`, context.uid)
      .count(),
  );

  if (userDisplayNamesSnap.data().count > 0)
    throw errors.exists(`User with given display name already exists`);
};

const uploadAvatar = async ({
  base64Avatar,
  context,
}: {
  context: ProtectedControllerHandlerContext;
  base64Avatar: Base64;
}): Promise<UserProfileModel['avatar']> => {
  const allowedExtensions = [`png`, `jpeg`, `webp`, `jpg`];
  const allowedMaxSize = 4;

  const avatar = decodeBase64Asset(base64Avatar);

  if (!allowedExtensions.includes(avatar.extension)) {
    throw errors.badRequest(
      `Invalid extension of avatar. Only ${allowedExtensions.join(
        `, `,
      )} are supported`,
    );
  }

  if (avatar.size > allowedMaxSize) {
    throw errors.badRequest(
      `Invalid avatar size. Maximum allowed is ${allowedMaxSize} MB`,
    );
  }

  const bucket = await storage().bucket();

  const rescalePromises: Promise<Buffer>[] = [];

  avatarVariants.forEach(({ h, w }) => {
    rescalePromises.push(
      sharp(avatar.buffer).resize(w, h).webp({ quality: 60 }).toBuffer(),
    );
  });

  const rescaleBuffers = await Promise.all(rescalePromises);
  const savePromises: Promise<void>[] = [];
  const paths: string[] = [];

  avatarVariants.forEach(({ size }, idx) => {
    const path = `${context.uid}/avatars/${size}`;
    const file = bucket.file(path);
    const buffer = rescaleBuffers[idx];

    savePromises.push(
      file.save(buffer, {
        contentType: `webp`,
      }),
    );
    paths.push(path);
  });

  await Promise.all(savePromises);

  return avatarVariants.reduce<NonNullable<UserProfileModel['avatar']>>(
    (acc, { size, w, h }, idx) => ({
      ...acc,
      [size]: {
        h,
        w,
        ext: `webp`,
        src: `https://firebasestorage.googleapis.com/v0/b/${
          bucket.name
        }/o/${encodeURIComponent(paths[idx])}?alt=media`,
        id: uuid(),
      },
    }),
    {} as NonNullable<UserProfileModel['avatar']>,
  );
};

const removeAvatar = async ({
  context,
}: {
  context: ProtectedControllerHandlerContext;
}) => {
  const bucket = await storage().bucket();

  await Promise.all(
    avatarVariants.map(({ size }) =>
      bucket.file(`${context.uid}/avatars/${size}`).delete(),
    ),
  );
};

const updateYourUserProfileHandler = async ({
  context,
  payload,
}: UpdateYourUserProfileHandlerConfig): Promise<UpdateYourUserProfileDto> => {
  const yourUserProfileRef = context.db
    .collection(`users-profiles`)
    .doc(context.uid);

  return await context.db.runTransaction(async (transaction) => {
    const [yourUserProfileSnap] = await Promise.all([
      transaction.get(yourUserProfileRef),
      verifyDisplayNamesDuplication({ transaction, payload, context }),
    ]);

    const userProfilePartial: Omit<
      UserProfileModel,
      'id' | 'cdate' | 'mdate' | 'avatar'
    > = {
      displayNameSlug:
        payload.profile.displayName !== null
          ? createSlug(payload.profile.displayName)
          : null,
      displayName: payload.profile.displayName,
      bio: payload.profile.bio,
      linkedInUrl: payload.profile.linkedInUrl,
      githubUrl: payload.profile.githubUrl,
      fbUrl: payload.profile.fbUrl,
      blogUrl: payload.profile.blogUrl,
      twitterUrl: payload.profile.twitterUrl,
    };
    const now = nowISO();

    const yourUserProfile = yourUserProfileSnap.data() as
      | UserProfileModel
      | undefined;

    if (!yourUserProfile) {
      const newUserProfile: UserProfileModel = {
        ...userProfilePartial,
        id: uuid(),
        cdate: now,
        mdate: now,
        avatar:
          payload.profile.avatar.type === `update`
            ? await uploadAvatar({
                base64Avatar: payload.profile.avatar.data,
                context,
              })
            : null,
      };

      await transaction.create(yourUserProfileRef, newUserProfile);

      return {
        mdate: newUserProfile.mdate,
        profile: newUserProfile,
      };
    }

    if (payload.mdate !== yourUserProfile.mdate) {
      throw errors.outOfDate(
        `Profile cannot be updated. Please refresh and try again (the profile may have been modified on another device or session)`,
      );
    }

    if (payload.profile.avatar.type === `remove`) {
      await removeAvatar({ context });
    }

    const updatedYourUserProfile: UserProfileModel = {
      ...userProfilePartial,
      id: yourUserProfile.id,
      cdate: yourUserProfile.cdate,
      mdate: now,
      avatar:
        payload.profile.avatar.type === `noop`
          ? yourUserProfile.avatar
          : payload.profile.avatar.type === `remove`
          ? null
          : await uploadAvatar({
              base64Avatar: payload.profile.avatar.data,
              context,
            }),
    };

    await transaction.update(yourUserProfileRef, updatedYourUserProfile);

    return {
      mdate: now,
      profile: updatedYourUserProfile,
    };
  });
};

export { updateYourUserProfileHandler };
