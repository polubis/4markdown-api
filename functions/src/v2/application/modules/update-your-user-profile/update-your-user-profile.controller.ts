import { protectedController } from '@utils/controller';
import { parse } from '@utils/parse';
import {
  type UpdateYourUserProfileDto,
  updateYourUserProfilePayloadSchema,
} from './update-your-user-profile.contract';
import { updateYourUserProfileHandler } from './update-your-user-profile.handler';

const updateYourUserProfileController =
  protectedController<UpdateYourUserProfileDto>(async (rawPayload, context) => {
    return await updateYourUserProfileHandler({
      context,
      payload: await parse(updateYourUserProfilePayloadSchema, rawPayload),
    });
  });

export { updateYourUserProfileController };
