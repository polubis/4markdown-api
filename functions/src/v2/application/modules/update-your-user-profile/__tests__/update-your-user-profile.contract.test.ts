import { updateYourUserProfilePayloadSchema } from '../update-your-user-profile.contract';

describe(`Update user profile contract works when`, () => {
  const mdate = `2025-01-21T10:04:27.719Z`;
  const payload = {
    mdate,
    profile: {
      displayName: `John Doe`,
      avatar: {
        type: `update`,
        data: `data:image/jpeg;base64,dGVzdGluZyBjb250ZW50`,
      },
      bio: `This is a valid biography that meets the minimum length requirement of twenty characters`,
      githubUrl: `https://github.com/johndoe`,
      fbUrl: `https://facebook.com/johndoe`,
      linkedInUrl: `https://linkedin.com/in/johndoe`,
      blogUrl: `https://blog.johndoe.com`,
      twitterUrl: `https://twitter.com/johndoe`,
    },
  };

  it(`it accepts a valid complete payload`, () => {
    expect(
      updateYourUserProfilePayloadSchema.safeParse(payload),
    ).toMatchSnapshot();
  });

  it(`it accepts nullish displayName`, () => {
    expect(
      updateYourUserProfilePayloadSchema.safeParse({
        mdate,
        profile: {
          ...payload.profile,
          displayName: null,
        },
      }),
    ).toMatchSnapshot();
  });

  it(`it accepts a valid payload with null optional fields`, () => {
    expect(
      updateYourUserProfilePayloadSchema.safeParse({
        mdate,
        profile: {
          displayName: null,
          avatar: { type: `noop` },
          bio: null,
          githubUrl: null,
          fbUrl: null,
          linkedInUrl: null,
          blogUrl: null,
          twitterUrl: null,
        },
      }),
    ).toMatchSnapshot();
  });

  const invalidPayloads = [
    {
      description: `required fields are missing`,
      payload: {},
    },
    {
      description: `the data is null`,
      payload: null,
    },
    {
      description: `mdate is invalid`,
      payload: {
        mdate: `not-a-date`,
        profile: {
          displayName: `John`,
          avatar: { type: `noop` },
          bio: `This is a valid biography that meets the minimum length requirement`,
        },
      },
    },
    {
      description: `displayName is too short`,
      payload: {
        mdate,
        profile: {
          displayName: `J`,
          avatar: { type: `noop` },
          bio: `This is a valid biography that meets the minimum length requirement`,
        },
      },
    },
    {
      description: `displayName is too long`,
      payload: {
        mdate,
        profile: {
          displayName: `This name is way too long to be valid`,
          avatar: { type: `noop` },
          bio: `This is a valid biography that meets the minimum length requirement`,
        },
      },
    },
    {
      description: `displayName is too long`,
      payload: {
        mdate,
        profile: {
          displayName: `This name is way too long to be valid`,
          avatar: { type: `noop` },
          bio: `This is a valid biography that meets the minimum length requirement`,
        },
      },
    },
    {
      description: `bio is too short`,
      payload: {
        mdate,
        profile: {
          displayName: `John`,
          avatar: { type: `noop` },
          bio: `Too short bio`,
        },
      },
    },
    {
      description: `avatar has invalid type`,
      payload: {
        mdate,
        profile: {
          displayName: `John`,
          avatar: { type: `invalid` },
          bio: `This is a valid biography that meets the minimum length requirement`,
        },
      },
    },
    {
      description: `avatar update missing base64 data`,
      payload: {
        mdate,
        profile: {
          displayName: `John`,
          avatar: { type: `update` },
          bio: `This is a valid biography that meets the minimum length requirement`,
        },
      },
    },
    {
      description: `URLs are invalid`,
      payload: {
        mdate,
        profile: {
          displayName: `John`,
          avatar: { type: `noop` },
          bio: `This is a valid biography that meets the minimum length requirement`,
          githubUrl: `not-a-url`,
          fbUrl: `also-not-a-url`,
        },
      },
    },
    {
      description: `URLs are invalid`,
      payload: {
        mdate,
        profile: {
          displayName: `John`,
          avatar: { type: `noop` },
          bio: `This is a valid biography that meets the minimum length requirement`,
          githubUrl: `not-a-url`,
          fbUrl: `also-not-a-url`,
        },
      },
    },
  ];

  test.each(invalidPayloads)(`it rejects the payload if %s`, ({ payload }) => {
    expect(() => updateYourUserProfilePayloadSchema.parse(payload)).toThrow();
    expect(
      updateYourUserProfilePayloadSchema.safeParse(payload),
    ).toMatchSnapshot();
  });
});
