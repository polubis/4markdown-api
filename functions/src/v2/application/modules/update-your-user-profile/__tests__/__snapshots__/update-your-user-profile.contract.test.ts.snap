// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update user profile contract works when it rejects the payload if { description: 'URLs are invalid', payload: [Object] } 1`] = `
{
  "error": [ZodError: [
  {
    "validation": "url",
    "code": "invalid_string",
    "message": "The GitHub URL format is invalid",
    "path": [
      "profile",
      "githubUrl"
    ]
  },
  {
    "validation": "url",
    "code": "invalid_string",
    "message": "The Facebook URL format is invalid",
    "path": [
      "profile",
      "fbUrl"
    ]
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "linkedInUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "blogUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "twitterUrl"
    ],
    "message": "Required"
  }
]],
  "success": false,
}
`;

exports[`Update user profile contract works when it rejects the payload if { description: 'avatar has invalid type', payload: [Object] } 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "received": "invalid",
            "code": "invalid_literal",
            "expected": "noop",
            "path": [
              "profile",
              "avatar",
              "type"
            ],
            "message": "Invalid literal value, expected \\"noop\\""
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "received": "invalid",
            "code": "invalid_literal",
            "expected": "update",
            "path": [
              "profile",
              "avatar",
              "type"
            ],
            "message": "Invalid literal value, expected \\"update\\""
          },
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [
              "profile",
              "avatar",
              "data"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "received": "invalid",
            "code": "invalid_literal",
            "expected": "remove",
            "path": [
              "profile",
              "avatar",
              "type"
            ],
            "message": "Invalid literal value, expected \\"remove\\""
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [
      "profile",
      "avatar"
    ],
    "message": "Invalid input"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "githubUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "fbUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "linkedInUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "blogUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "twitterUrl"
    ],
    "message": "Required"
  }
]],
  "success": false,
}
`;

exports[`Update user profile contract works when it rejects the payload if { description: 'avatar update missing base64 data', payload: [Object] } 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "received": "update",
            "code": "invalid_literal",
            "expected": "noop",
            "path": [
              "profile",
              "avatar",
              "type"
            ],
            "message": "Invalid literal value, expected \\"noop\\""
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [
              "profile",
              "avatar",
              "data"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "received": "update",
            "code": "invalid_literal",
            "expected": "remove",
            "path": [
              "profile",
              "avatar",
              "type"
            ],
            "message": "Invalid literal value, expected \\"remove\\""
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [
      "profile",
      "avatar"
    ],
    "message": "Invalid input"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "githubUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "fbUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "linkedInUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "blogUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "twitterUrl"
    ],
    "message": "Required"
  }
]],
  "success": false,
}
`;

exports[`Update user profile contract works when it rejects the payload if { description: 'bio is too short', payload: [Object] } 1`] = `
{
  "error": [ZodError: [
  {
    "code": "too_small",
    "minimum": 20,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "Bio must be at least 20 characters long",
    "path": [
      "profile",
      "bio"
    ]
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "githubUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "fbUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "linkedInUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "blogUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "twitterUrl"
    ],
    "message": "Required"
  }
]],
  "success": false,
}
`;

exports[`Update user profile contract works when it rejects the payload if { description: 'displayName is too long', payload: [Object] } 1`] = `
{
  "error": [ZodError: [
  {
    "code": "too_big",
    "maximum": 30,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "Display name can be up to 30 characters long",
    "path": [
      "profile",
      "displayName"
    ]
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "githubUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "fbUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "linkedInUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "blogUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "twitterUrl"
    ],
    "message": "Required"
  }
]],
  "success": false,
}
`;

exports[`Update user profile contract works when it rejects the payload if { description: 'displayName is too long', payload: [Object] } 2`] = `
{
  "error": [ZodError: [
  {
    "code": "too_big",
    "maximum": 30,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "Display name can be up to 30 characters long",
    "path": [
      "profile",
      "displayName"
    ]
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "githubUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "fbUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "linkedInUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "blogUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "twitterUrl"
    ],
    "message": "Required"
  }
]],
  "success": false,
}
`;

exports[`Update user profile contract works when it rejects the payload if { description: 'displayName is too short', payload: [Object] } 1`] = `
{
  "error": [ZodError: [
  {
    "code": "too_small",
    "minimum": 2,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "Display name must be at least 2 characters long",
    "path": [
      "profile",
      "displayName"
    ]
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "githubUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "fbUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "linkedInUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "blogUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "twitterUrl"
    ],
    "message": "Required"
  }
]],
  "success": false,
}
`;

exports[`Update user profile contract works when it rejects the payload if { description: 'mdate is invalid', payload: [Object] } 1`] = `
{
  "error": [ZodError: [
  {
    "validation": "regex",
    "code": "invalid_string",
    "message": "Invalid",
    "path": [
      "mdate"
    ]
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "githubUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "fbUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "linkedInUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "blogUrl"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "profile",
      "twitterUrl"
    ],
    "message": "Required"
  }
]],
  "success": false,
}
`;

exports[`Update user profile contract works when it rejects the payload if { description: 'required fields are missing', payload: {} } 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "mdate"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "undefined",
    "path": [
      "profile"
    ],
    "message": "Required"
  }
]],
  "success": false,
}
`;

exports[`Update user profile contract works when it rejects the payload if { description: 'the data is null', payload: null } 1`] = `
{
  "error": [ZodError: [
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "null",
    "path": [],
    "message": "Expected object, received null"
  }
]],
  "success": false,
}
`;
